app/src/swap.c

```diff
@@ -0,0 +1,249 @@
+    parser_error_t err = parser_unexpected_error;
+    if (ctx == NULL) {
+        return err;
+    }
+    // Check method.
+    const char * valid_tx_method = "Balances Transfer";
+    char tmp_str[80] = {0};
+    snprintf(tmp_str, sizeof(tmp_str), "%s %s", _getMethod_ModuleName(ctx->tx_obj->transactionVersion,
+                                                                    ctx->tx_obj->callIndex.moduleIdx),
+                                                _getMethod_Name(ctx->tx_obj->transactionVersion,
+                                                                ctx->tx_obj->callIndex.moduleIdx,
+                                                                ctx->tx_obj->callIndex.idx));
+
+    if (strncmp(tmp_str, &valid_tx_method[0], strlen(valid_tx_method)) != 0) {
+         PRINTF("Wrong swap tx method (%s, should be : %s).\n",tmp_str,valid_tx_method);
+         ZEMU_LOGF(100, "Wrong swap tx method (%s, should be : %s).\n",tmp_str,valid_tx_method)
+         return parser_swap_tx_wrong_method;
+    }
+
+    // Check transaction method arguments number. Balance transfer Should be 2 (for tx v18).
+    if (_getMethod_NumItems(ctx->tx_obj->transactionVersion,
+                           ctx->tx_obj->callIndex.moduleIdx,
+                           ctx->tx_obj->callIndex.idx) != 2) {
+        ZEMU_LOGF(50, "Wrong swap tx method arguments count.\n")
+        return parser_swap_tx_wrong_method_args_num;
+    }
+
+    // Check destination address.
+    MEMZERO(tmp_str,sizeof(tmp_str));
+    uint8_t pageCount = 0;
+    err = _getMethod_ItemValue(ctx->tx_obj->transactionVersion,
+                            &ctx->tx_obj->method,
+                            ctx->tx_obj->callIndex.moduleIdx, ctx->tx_obj->callIndex.idx, 0,
+                            tmp_str, sizeof(tmp_str),
+                            0, &pageCount);
+
+    if (err != parser_ok) {
+        ZEMU_LOGF(50, "Could not parse swap tx destination address.");
+        return err;
+    }
+
+    if (strncmp(tmp_str, &(G_swap_state.destination_address[0]), sizeof(G_swap_state.destination_address)) != 0) {
+         ZEMU_LOGF(100, "Wrong swap tx destination address (%s, should be : %s).\n", tmp_str, G_swap_state.destination_address)
+         return parser_swap_tx_wrong_dest_addr;
+    }
+
+    // Check amount.
+    MEMZERO(tmp_str, sizeof(tmp_str));
+    err = _getMethod_ItemValue(ctx->tx_obj->transactionVersion,
+                            &ctx->tx_obj->method,
+                            ctx->tx_obj->callIndex.moduleIdx, ctx->tx_obj->callIndex.idx, 1,
+                            tmp_str, sizeof(tmp_str),
+                            0, &pageCount);
+
+    if(err != parser_ok)
+    {
+        ZEMU_LOGF(50, "Could not parse swap tx amount.");
+        return err;
+    }
+
+    char tmp_amount[50] = {0};
+    pd_CompactBalance_t amount = {.value.ptr = G_swap_state.amount, .value.len = G_swap_state.amount_length};
```

Comment:
You never assign `G_swap_state.amount_length` in `copy_transaction_parameters`. You should add the line : 

`G_swap_state.amount_length = sign_transaction_params->amount_length;`

Reply:
Fixed in 25edf77bd70d1975aa2d29dd98049e25a6f99a2f